cpp/src/io/parquet/page_enc.cu

```diff
@@ -1663,38 +1743,60 @@ dremel_data get_dremel_data(column_view h_col,
     return std::make_tuple(std::move(empties), std::move(empties_idx), empties_size);
   };
 
-  // Reverse the nesting in order to merge the deepest level with the leaf first and merge bottom
-  // up
-  auto curr_col        = h_col;
-  size_t max_vals_size = 0;
+  auto curr_col = h_col;
   std::vector<column_view> nesting_levels;
   std::vector<uint8_t> def_at_level;
-  size_type level       = 0;
-  auto add_def_at_level = [&](size_type level) {
-    auto is_level_nullable =
-      curr_col.nullable() or (not level_nullability.empty() and level_nullability[level]);
-    def_at_level.push_back(is_level_nullable ? 2 : 1);
+  std::vector<uint8_t> start_at_sub_level;
+  uint8_t curr_nesting_level_idx = 0;
+
+  auto add_def_at_level = [&](column_view col) {
+    // Add up all def level contributions in this column all the way till the first list column
+    // appears in the hierarchy or until we get to leaf
+    uint32_t def = 0;
+    start_at_sub_level.push_back(curr_nesting_level_idx);
+    while (col.type().id() == type_id::STRUCT) {
+      def += (nullability[curr_nesting_level_idx]) ? 1 : 0;
+      col = col.child(0);
+      ++curr_nesting_level_idx;
+    }
+    // At the end of all those structs is either a list column or the leaf. Leaf column contributes
+    // at least one def level. It doesn't matter what the leaf contributes because it'll be at the
+    // end of the exclusive scan.
+    def += (nullability[curr_nesting_level_idx]) ? 2 : 1;
+    def_at_level.push_back(def);
+    ++curr_nesting_level_idx;
   };
-  while (curr_col.type().id() == type_id::LIST) {
+  while (cudf::is_nested(curr_col.type())) {
     nesting_levels.push_back(curr_col);
-    add_def_at_level(level);
-    auto lcv = lists_column_view(curr_col);
-    max_vals_size += lcv.offsets().size();
-    curr_col = lcv.child();
-    level++;
+    add_def_at_level(curr_col);
+    while (curr_col.type().id() == type_id::STRUCT) {
+      // Go down the hierarchy until we get to the LIST or the leaf level
+      curr_col = curr_col.child(0);
+    }
+    if (curr_col.type().id() == type_id::LIST) {
+      curr_col = curr_col.child(lists_column_view::child_column_index);
+      if (not is_nested(curr_col.type())) {
+        // Special case: when the leaf data column is the immediate child of the list col then we
+        // want it to be included right away. Otherwise the struct containing it will be included in
+        // the next iteration of this loop.
+        nesting_levels.push_back(curr_col);
+        add_def_at_level(curr_col);
+        break;
+      }
+    }
```

Comment:
This one's really something.
Looking at the code it seems like struct(list(int)) would add def levels for struct and int columns. In contrast, list(struct(int)) would add def levels for list and struct columns. Is this the intended behavior?

Reply:
That is the intended behaviour. Here's how this works:
At each list level we want to do a thrust::merge of the definition levels between the parent and the child. In case of struct columns, we don't have to merge struct levels with their children because a struct is the same size as its children. e.g. for a column `struct<int, float>`, if the row `i` is null, then the children columns `int` and `float` are also null at `i`. They also have the null entry represented in their respective null masks. So for any case of strictly struct based nesting, we can get the definition levels merely by iterating over the nesting for the same row.

In case struct and lists are intermixed, the definition levels of all the contiguous struct levels can be constructed using the aforementioned iterative method. Only when we reach a list level, we need to do a merge with the subsequent level.

So, for a column like `struct<list<int>>`, we are going to merge between the levels `struct<list` and `int`.
For a column like `list<struct<int>>`, we are going to merge between `list` and `struct<int>`.

In general, one nesting level as the list level and any struct level that precedes it.
A few more examples I have in my notebook to make it easier to visualize: (L is list, S is struct, angle brackets omitted)
```
1. LSi     = L   Si
             - | --

2. LLSi    = L   L   Si
             - | - | --

3. SSLi    = SSL   i
             --- | -

4. LLSLSSi = L   L   SL   SSi
             - | - | -- | ---
```

Reply:
Expanded doxygen docs.